opengrok-indexer/src/main/java/org/opengrok/indexer/analysis/Ctags.java

```diff
@@ -390,7 +398,35 @@ public Definitions doCtags(String file) throws IOException,
             if (Thread.interrupted()) {
                 throw new InterruptedException("flush()");
             }
+
+            int timeout = this.timeout * 1000;
+            Timer timer = null;
+            /*
+             * Setup timer thread to make sure the ctags process completes and the indexer can
+             * make progress instead of hanging the whole operation.
+             */
+            if (timeout != 0) {
+                timer = new Timer();
+                timer.schedule(new TimerTask() {
+                    @Override public void run() {
+                        LOGGER.log(Level.WARNING,
+                                String.format("Terminating ctags process for file '%s' " +
+                                        "due to timeout %d seconds", file, timeout / 1000));
+                        try {
+                            close();
+                        } catch (IOException e) {
+                            LOGGER.log(Level.WARNING, "Failed to terminate overly long ctags command");
+                        }
+                    }
+                }, timeout);
+            }
+
             readTags(rdr);
+
+            if (timer != null) {
```

Comment:
Since you are not using any synchronization, then it is theoretically possible that `readTags()` would be successful and you would still log warning message and close the ctags prematurely, right? I know that this is an edge case with the timeout so maybe we can waive it?

Reply:
true, there is a time window between when `readTags()` finishes and the `timer` check.

Reply:
the `Future` based implementation deals with that.